packages/protocol/src/extension.js

```diff
@@ -0,0 +1,310 @@
+//
+// Copyright 2019 DxOS.
+//
+
+import assert from 'assert';
+import { EventEmitter } from 'events';
+import crypto from 'crypto';
+
+import debug from 'debug';
+
+import { keyToHuman } from './utils';
+import { Codec } from './codec';
+import { ProtocolError } from './protocol';
+
+const log = debug('extension');
+
+/**
+ * Reliable message passing via using Dat protocol extensions.
+ *
+ * Events: "send", "receive", "error"
+ */
+export class Extension extends EventEmitter {
+  /**
+   * @type {Protocol}
+   */
+  _protocol = null;
+
+  /**
+   * Pending messages.
+   * @type {Map<{id, Function}>}
+   */
+  _pendingMessages = new Map();
+
+  /**
+   * Handshake handler.
+   * @type {Function<{protocol, context}>}
+   */
+  _handshakeHandler = null;
+
+  /**
+   * Close handler.
+   * @type {Function<{protocol, context}>}
+   */
+  _closeHandler = null;
+
+  /**
+   * Message handler.
+   * @type {Function<{protocol, context, message}>}
+   */
+  _messageHandler = null;
+
+  /**
+   * Feed handler.
+   * @type {Function<{protocol, context, discoveryKey}>}
+   */
+  _feedHandler = null;
+
+  _stats = {
+    send: 0,
+    receive: 0,
+    error: 0
+  };
+
+  /**
+   * @param {string} name
+   * @param {Object} options
+   * @param {Number} options.timeout
+   * @param {Codec} options.codec
+   */
+  constructor (name, options = {}) {
+    super();
+    assert(typeof name === 'string' && name.length > 0, 'Name is required.');
+
+    this._name = name;
+
+    this._options = Object.assign({
+      timeout: 2000,
+      codec: new Codec({ binary: options.binary })
+    }, options);
+  }
+
+  get name () {
+    return this._name;
+  }
+
+  get stats () {
+    return this._stats;
+  }
+
+  /**
+   * Sets the handshake handler.
+   * @param {Function<{protocol, context}>} handshakeHandler - Async handshake handler.
+   * @returns {Extension}
+   */
+  setHandshakeHandler (handshakeHandler) {
```

Comment:
Should we prefer events here (I like callbacks -- and this may have been me in a previous life :) -- but are events more consistent)? Makes emit less code than checking the callback.

Reply:
We are not going to refactor the protocol in this PR, we are moving the current code from wireline-core to dxos.